### PR TITLE
Enum flattening

### LINKED ISSUE #96: Unable to generate flattened enum
I have tried both tagged or untagged enum in struct, not working.

## Code

Tagged:

```rust
use serde::Serialize;
use ts_rs::TS;

#[derive(Serialize, TS)]
#[ts(export)]
struct Foo {
    #[serde(flatten)]
    baz: Bar,
}

#[derive(Serialize, TS)]
#[serde(tag = "tag")]
#[ts(export)]
enum Bar {
    _Baz { a: i32 },
}
```

Untagged:

```rust
use serde::Serialize;
use ts_rs::TS;

#[derive(Serialize, TS)]
#[ts(export)]
struct Foo {
    #[serde(flatten)]
    baz: Bar,
}

#[derive(Serialize, TS)]
#[serde(untagged)]
#[ts(export)]
enum Bar {
    _Baz,
}
```

## Error 

```
---- ts::export_bindings_foo stdout ----
thread 'ts::export_bindings_foo' panicked at 'Bar cannot be flattened', [:REDACTED:]/.cargo/registry/src/github.com-1ecc6299db9ec823/ts-rs-6.1.2/src/lib.rs:243:9
```

## Info
version of `ts-rs`: `6.1.2`
version of `rustc`: `rustc 1.60.0 (7737e0b5c 2022-04-04)`

### PR DESCRIPTION
**Pull request by escritorio-gustavo**
Refactored version of #202 created as a separate PR because their histories are not compatible

Fixes #96

### CODE REVIEW

**Comment by NyxCode** [comment]
Amazing! The previous PR was super hard to understand (for me, at least :D). This here is now very concise, absolutely love it!

**Comment by escritorio-gustavo** [comment]
It really was, I was still trying to to figure out how to generate `inline_flattened` for enums and it hadn't occurred to me that I could also do type intersections for structs

File: macros/src/types/enum.rs [hunk 1]
```diff
@@ -130,7 +134,7 @@ fn format_variant(
                     "{{ \"{}\": \"{}\", {} }}",
                     #tag,
                     #name,
-                    #inline_flattened
+                    #inline_flattened.trim_matches(&['{', '}', ' '])
```
Comment (NyxCode) [inline macros/src/types/enum.rs:137]: I think here, a comment would be usefull - How does `inline_flattened` look like, and why do we need to trim the braces here?

File: macros/src/types/named.rs [hunk 2]
```diff
@@ -57,6 +63,7 @@ pub(crate) fn named(
 // build an expresion which expands to a string, representing a single field of a struct.
 fn format_field(
```
Comment (NyxCode) [inline macros/src/types/named.rs:64]: Same here. Maybe just 1 sentence explaining when the expression ends up in flattened_fields and when in formatted_fields, and why these two vecs are required.

File: ts-rs/tests/enum_flattening.rs [hunk 3]
```diff
@@ -0,0 +1,104 @@
+use serde::Serialize;
```
Comment (NyxCode) [inline ts-rs/tests/enum_flattening.rs:1]: We're getting some unused import warnings here.  
Also, we're not testing `#[ts(flatten)]`, which should be equivalent do `#[serde(flatten)]` with compat-mode on. 
I think with a few `#[cfg(..)]`s and `#[cfg_attr(..)]`s, we can make the test work with both compat-mode on and off as-is.

**Review by NyxCode — CHANGES_REQUESTED** [review]

File: ts-rs/tests/enum_flattening.rs (hunk 3 above)
Comment (escritorio-gustavo) [reply ts-rs/tests/enum_flattening.rs:1]: Yeah, I'm adding those now, but the only test that will work without `serde-compat` is `internally_tagged` as all the others require `#[serde(tag = "...")]` or `#[serde(untagged)]`

File: macros/src/types/named.rs (hunk 2 above)
Comment (escritorio-gustavo) [reply macros/src/types/named.rs:73]: Done!

File: macros/src/types/enum.rs (hunk 1 above)
Comment (escritorio-gustavo) [reply macros/src/types/enum.rs:145]: Done!

Comment (escritorio-gustavo) [reply macros/src/types/enum.rs:145]: Basically, using [serde's example](https://serde.rs/enum-representations.html#internally-tagged), a type that should turn into this:
```ts
{"type": "Request", id: string, method: string, params: { /*...*/ } }
```

would turn into this
```ts
{"type": "Request", { id: string, method: string, params: { /*...*/ } } }
```

**Review by NyxCode — APPROVED** [review]

**Comment by NyxCode** [comment]
Awesomee, thanks for the comments! From my side, I think this is ready for a merge.
Great work, i'm very happy to see support for this. Also, thanks for cleaning up the jungle of open issues & PRs!

**Pull request merged**